### PR TITLE
Add nomination pools to Polkadot runtime

### PR DESCRIPTION
**Pull request by kianenigma**
- targeting release 0.9.23

- Also sets the initial configs of the pools: 

- 5 DOT to join a pool.
- 100 DOT to create a pool.
- 64 initial pools: only for initial safety: can be set to infinity when needed.
- 4096 members per pool: only for initial safety: can be set to infinity when needed.
- 128k total pool members: only for initial safety: can be set to infinity when needed.

Should only be merged once the following is merged on substrate: 

- [x] https://github.com/paritytech/substrate/pull/11426
- [x] https://github.com/paritytech/substrate/pull/11669

## UPDATE:

The above numbers are still my suggestion, but I this PR as it is has all the max counters set to 0, meaning that we can deploy this but no pool can be created. This will be helpful in making the process of enabling pools independent of a runtime upgrade.

### CODE REVIEW

**Comment by ggwpez** [comment]
I think you currently have to copy&paste the benchmark command without the `--heap-pages` flag to make it work like here:  https://github.com/paritytech/substrate/pull/11431#issuecomment-1133018018.
Should be fixed after: https://github.com/paritytech/substrate/pull/11482

**Comment by ggwpez** [comment]
/benchmark runtime custom pallet --chain=polkadot-dev --steps=50 --repeat=20 --pallet=pallet_nomination_pools --extrinsic="*" --execution=wasm --wasm-execution=compiled --header=./file_header.txt --output=./runtime/polkadot/src/weights/

**Review by shawntabrizi — COMMENTED** [review]
When does this plan to go out?

I feel we should probably let this bake on Kusama a bit longer, especially given the recent bug reports.

**Comment by bkchr** [comment]
> When does this plan to go out?
> 
> I feel we should probably let this bake on Kusama a bit longer, especially given the recent bug reports.

Sounds like a good plan to me.

**Comment by kianenigma** [comment]
Given

> The above numbers are still my suggestion, but I this PR as it is has all the max counters set to 0, meaning that we can deploy this but no pool can be created. This will be helpful in making the process of enabling pools independent of a runtime upgrade.

This is ready to be merged.

**Comment by ggwpez** [comment]
/cmd queue -c bench-bot $ runtime polkadot-dev pallet_nomination_pools

**Comment by kianenigma** [comment]
bot rebase

File: runtime/polkadot/src/lib.rs [hunk 1]
```diff
@@ -1389,6 +1391,52 @@ impl auctions::Config for Runtime {
 	type WeightInfo = weights::runtime_common_auctions::WeightInfo<Runtime>;
 }
 
+parameter_types! {
+	pub const PoolsPalletId: PalletId = PalletId(*b"py/nopls");
+	pub const MaxPointsToBalance: u8 = 10;
```
Comment (ggwpez) [inline runtime/polkadot/src/lib.rs:1396]:
```suggestion
	// Allow pools that got slashed up to 90% to remain operational.
	pub const MaxPointsToBalance: u8 = 10;
```

File: runtime/polkadot/src/lib.rs [hunk 2]
```diff
@@ -1389,6 +1391,52 @@ impl auctions::Config for Runtime {
...
+			// 100 DOT to create a pool.
+			pallet_nomination_pools::MinCreateBond::<Runtime>::put(100 * UNITS);
+
+			// Initialize with limits for now.
+			pallet_nomination_pools::MaxPools::<Runtime>::put(0);
+			pallet_nomination_pools::MaxPoolMembersPerPool::<Runtime>::put(0);
+			pallet_nomination_pools::MaxPoolMembers::<Runtime>::put(0);
+
+			log::info!(target: "runtime::polkadot", "pools config initiated 🎉");
+			<Runtime as frame_system::Config>::DbWeight::get().reads_writes(1, 5)
```
Comment (ggwpez) [inline runtime/polkadot/src/lib.rs:1432]: You could call the `sanity_check` here. Just an idea.

File: runtime/polkadot/src/lib.rs [hunk 3]
```diff
@@ -1389,6 +1391,52 @@ impl auctions::Config for Runtime {
...
+	type PostUnbondingPoolsWindow = frame_support::traits::ConstU32<4>;
+	type MaxMetadataLen = frame_support::traits::ConstU32<256>;
+	// we use the same number of allowed unlocking chunks as with staking.
+	type MaxUnbonding = <Self as pallet_staking::Config>::MaxUnlockingChunks;
+	type PalletId = PoolsPalletId;
+	type MaxPointsToBalance = MaxPointsToBalance;
+	type WeightInfo = weights::pallet_nomination_pools::WeightInfo<Self>;
+}
+
+pub struct InitiatePoolConfigs;
```
Comment (ggwpez) [inline runtime/polkadot/src/lib.rs:1416]: Does this now need to be added to the `Executive` to make it execute or will the release team do that?

File: runtime/common/src/lib.rs [hunk 4]
```diff
@@ -215,6 +215,23 @@ impl pallet_staking::BenchmarkingConfig for StakingBenchmarkingConfig {
 	type MaxNominators = ConstU32<1000>;
 }
 
+/// Convert a balance to an unsigned 256-bit number, use in nomination pools.
+pub struct BalanceToU256;
+impl sp_runtime::traits::Convert<Balance, sp_core::U256> for BalanceToU256 {
+	fn convert(n: Balance) -> sp_core::U256 {
+		n.into()
```
Comment (ggwpez) [inline runtime/common/src/lib.rs:222]: Why is this not using `try_into` and defensive like below?

**Review by ggwpez — APPROVED** [review]
Exciting! Lets go :rocket:

File: runtime/common/src/lib.rs (hunk 4 above)
Comment (kianenigma) [reply runtime/common/src/lib.rs:222]: u128 can infallibly be converted to u128, there is no `try_into` here.

File: runtime/polkadot/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply runtime/polkadot/src/lib.rs:1433]: What should I do if it fails? :)) probably best to put it in post-upgrade.

File: runtime/polkadot/src/lib.rs (hunk 3 above)
Comment (kianenigma) [reply runtime/polkadot/src/lib.rs:1417]: ah damn, it needs to be added.

File: runtime/polkadot/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply runtime/polkadot/src/lib.rs:1433]: Unfortunately they are not public at the moment.

**Review by shawntabrizi — APPROVED** [review]

File: runtime/polkadot/src/lib.rs [hunk 5]
```diff
@@ -1389,6 +1391,53 @@ impl auctions::Config for Runtime {
...
+	type MaxMetadataLen = frame_support::traits::ConstU32<256>;
+	// we use the same number of allowed unlocking chunks as with staking.
+	type MaxUnbonding = <Self as pallet_staking::Config>::MaxUnlockingChunks;
+	type PalletId = PoolsPalletId;
+	type MaxPointsToBalance = MaxPointsToBalance;
+	type WeightInfo = weights::pallet_nomination_pools::WeightInfo<Self>;
+}
+
+pub struct InitiateNominationPools;
+impl frame_support::traits::OnRuntimeUpgrade for InitiateNominationPools {
```
Comment (shawntabrizi) [inline runtime/polkadot/src/lib.rs:1418]: in the future probably would be nice to have an initialize function coming from the pallet itself.

File: runtime/polkadot/src/lib.rs [hunk 6]
```diff
@@ -1389,6 +1391,53 @@ impl auctions::Config for Runtime {
...
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// we use one as an indicator if this has already been set.
+		if pallet_nomination_pools::MaxPools::<Runtime>::get().is_none() {
+			// 1 DOT to join a pool.
```
Comment (bkchr) [inline runtime/polkadot/src/lib.rs:1422]: Comment and code are not matching 😏😏

**Review by bkchr — APPROVED** [review]

File: runtime/polkadot/src/lib.rs (hunk 6 above)
Comment (kianenigma) [inline runtime/polkadot/src/lib.rs:1422]:
```suggestion
			// 5 DOT to join a pool.
```

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**